
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Welcome to pysradb’s documentation! &#8212; pysradb 0.10.5-dev0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/graphviz.css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Quickstart" href="quickstart.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <img alt="_images/pysradb_v3.png" src="_images/pysradb_v3.png" />
<div class="line-block">
<div class="line"><br /></div>
</div>
<div class="section" id="welcome-to-pysradb-s-documentation">
<h1>Welcome to pysradb’s documentation!<a class="headerlink" href="#welcome-to-pysradb-s-documentation" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>The NCBI Sequence Read Archive (SRA) is the primary archive of
next-generation sequencing datasets. SRA makes metadata and raw sequencing
data available to the research community to encourage reproducibility and
to provide avenues for testing novel hypotheses on publicly available data.</p>
<p><code class="docutils literal notranslate"><span class="pre">pysradb</span></code> provides a simple method to programmatically access metadata
and download sequencing data from SRA and European Bioinformatics
Institute’s European Reads Archive (ENA).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$</span> pysradb metadata SRP265425
<span class="go">study_accession experiment_accession    experiment_title        experiment_desc organism_taxid  organism_name   library_strategy        library_source  library_selection       sample_accession        sample_title    instrument      total_spots     total_size      run_accession   run_total_spots run_total_bases</span>
<span class="go">SRP265425       SRX8434255      Ampliseq of SARS-CoV-2  Ampliseq of SARS-CoV-2  2697049 Severe acute respiratory syndrome coronavirus 2 AMPLICON        VIRAL RNA       RT-PCR  SRS6745319      Ion Torrent S5 XL       1311358 83306910       SRR11886735      1311358 109594216</span>
<span class="go">SRP265425       SRX8434254      Ampliseq of SARS-CoV-2  Ampliseq of SARS-CoV-2  2697049 Severe acute respiratory syndrome coronavirus 2 AMPLICON        VIRAL RNA       RT-PCR  SRS6745320      Ion Torrent S5 XL       2614109 204278682      SRR11886736      2614109 262305651</span>
<span class="go">SRP265425       SRX8434253      Ampliseq of SARS-CoV-2  Ampliseq of SARS-CoV-2  2697049 Severe acute respiratory syndrome coronavirus 2 AMPLICON        VIRAL RNA       RT-PCR  SRS6745318      Ion Torrent S5 XL       2286312 183516004      SRR11886737      2286312 263304134</span>
<span class="go">SRP265425       SRX8434252      Ampliseq of SARS-CoV-2  Ampliseq of SARS-CoV-2  2697049 Severe acute respiratory syndrome coronavirus 2 AMPLICON        VIRAL RNA       RT-PCR  SRS6745317      Ion Torrent S5 XL       5202567 507524965      SRR11886738      5202567 781291588</span>
<span class="go">SRP265425       SRX8434251      Ampliseq of SARS-CoV-2  Ampliseq of SARS-CoV-2  2697049 Severe acute respiratory syndrome coronavirus 2 AMPLICON        VIRAL RNA       RT-PCR  SRS6745315      Ion Torrent S5 XL       3313960 356104406      SRR11886739      3313960 612430817</span>
<span class="go">SRP265425       SRX8434250      Ampliseq of SARS-CoV-2  Ampliseq of SARS-CoV-2  2697049 Severe acute respiratory syndrome coronavirus 2 AMPLICON        VIRAL RNA       RT-PCR  SRS6745316      Ion Torrent S5 XL       5155733 565882351      SRR11886740      5155733 954342917</span>
<span class="go">SRP265425       SRX8434249      Ampliseq of SARS-CoV-2  Ampliseq of SARS-CoV-2  2697049 Severe acute respiratory syndrome coronavirus 2 AMPLICON        VIRAL RNA       RT-PCR  SRS6745313      Ion Torrent S5 XL       1324589 175619046      SRR11886741      1324589 216531400</span>
<span class="go">SRP265425       SRX8434248      Ampliseq of SARS-CoV-2  Ampliseq of SARS-CoV-2  2697049 Severe acute respiratory syndrome coronavirus 2 AMPLICON        VIRAL RNA       RT-PCR  SRS6745314      Ion Torrent S5 XL       1639851 198973268      SRR11886742      1639851 245466005</span>
<span class="go">SRP265425       SRX8434247      Ampliseq of SARS-CoV-2  Ampliseq of SARS-CoV-2  2697049 Severe acute respiratory syndrome coronavirus 2 AMPLICON        VIRAL RNA       RT-PCR  SRS6745312      Ion Torrent S5 XL       3921389 210198580      SRR11886743      3921389 332935558</span>
<span class="go">SRP265425       SRX8434246      Ampliseq of SARS-CoV-2  Ampliseq of SARS-CoV-2  2697049 Severe acute respiratory syndrome coronavirus 2 AMPLICON        VIRAL RNA       RT-PCR  SRS6745311      Ion Torrent S5 XL       14295475        2150005008      SRR11886744     14295475        2967829315</span>
<span class="go">SRP265425       SRX8434245      Ampliseq of SARS-CoV-2  Ampliseq of SARS-CoV-2  2697049 Severe acute respiratory syndrome coronavirus 2 AMPLICON        VIRAL RNA       RT-PCR  SRS6745310      Ion Torrent S5 XL       5124692 294846140      SRR11886745      5124692 431819462</span>
<span class="go">SRP265425       SRX8434244      Ampliseq of SARS-CoV-2  Ampliseq of SARS-CoV-2  2697049 Severe acute respiratory syndrome coronavirus 2 AMPLICON        VIRAL RNA       RT-PCR  SRS6745309      Ion Torrent S5 XL       2986306 205666872      SRR11886746      2986306 275400959</span>
<span class="go">SRP265425       SRX8434243      Ampliseq of SARS-CoV-2  Ampliseq of SARS-CoV-2  2697049 Severe acute respiratory syndrome coronavirus 2 AMPLICON        VIRAL RNA       RT-PCR  SRS6745308      Ion Torrent S5 XL       1182690 59471336       SRR11886747      1182690 86350631</span>
<span class="go">SRP265425       SRX8434242      Ampliseq of SARS-CoV-2  Ampliseq of SARS-CoV-2  2697049 Severe acute respiratory syndrome coronavirus 2 AMPLICON        VIRAL RNA       RT-PCR  SRS6745307      Ion Torrent S5 XL       6031816 749323230      SRR11886748      6031816 928054297</span>
</pre></div>
</div>
<p>The strength of <code class="docutils literal notranslate"><span class="pre">pysradb</span></code> lies in its versatility: It can be used as
either a command-line application to easily retrieve the desired
information. It can also be used as a python package, which organises
metadata as <cite>pandas</cite> DataFrames, providing more control for data
exploration.</p>
</div>
<hr class="docutils" />
<div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<div class="section" id="stable-release">
<h3>Stable release<a class="headerlink" href="#stable-release" title="Permalink to this headline">¶</a></h3>
<p>To install pysradb, run this command in your terminal:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$</span> pip install pysradb
</pre></div>
</div>
<p>This is the preferred method to install pysradb, as it will always install the most recent stable release.</p>
<p>If you don’t have <a class="reference external" href="https://pip.pypa.io">pip</a> installed, this <a class="reference external" href="http://docs.python-guide.org/en/latest/starting/installation/">Python installation guide</a> can guide
you through the process.</p>
<p>Alternatively, you may use conda:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda install -c bioconda pysradb
</pre></div>
</div>
<p>This step will install all the dependencies except aspera-client (which is not required, but highly recommended).
If you have an existing environment with a lot of pre-installed packages, conda might be <a class="reference external" href="https://github.com/bioconda/bioconda-recipes/issues/13774">slow</a>.
Please consider creating a new enviroment for <code class="docutils literal notranslate"><span class="pre">pysradb</span></code>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda create -c bioconda -n pysradb <span class="nv">PYTHON</span><span class="o">=</span><span class="m">3</span> pysradb
</pre></div>
</div>
</div>
<div class="section" id="from-sources">
<h3>From sources<a class="headerlink" href="#from-sources" title="Permalink to this headline">¶</a></h3>
<p>The source files for pysradb can be downloaded from the <a class="reference external" href="https://github.com/saketkc/pysradb">Github repo</a>.</p>
<p>You can either clone the public repository:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$</span> git clone git://github.com/saketkc/pysradb
</pre></div>
</div>
<p>Or download the <a class="reference external" href="https://github.com/saketkc/pysradb/tarball/master">tarball</a>:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$</span> curl  -OL https://github.com/saketkc/pysradb/tarball/master
</pre></div>
</div>
<p>Once you have a copy of the source, you can install it with:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$</span> python setup.py install
</pre></div>
</div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="using-pysradb">
<h2>Using pysradb<a class="headerlink" href="#using-pysradb" title="Permalink to this headline">¶</a></h2>
<p>See <a class="reference internal" href="quickstart.html"><span class="doc">Quickstart</span></a> to quickly get started using pysradb</p>
<p>Please see <a class="reference internal" href="usage_scenarios.html"><span class="doc">Usage Scenarios</span></a> for a few usage scenarios with detailed explanation.</p>
</div>
<hr class="docutils" />
<div class="section" id="contents">
<h2>Contents<a class="headerlink" href="#contents" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a><ul>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#install-pysradb">Install pysradb</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#using-pysradb">Using pysradb</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#the-full-list-of-possible-pysradb-operations">The full list of possible <cite>pysradb</cite> operations</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="cmdline.html">CLI Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="cmdline.html#getting-metadata-for-a-sra-project-srp">Getting metadata for a SRA project (SRP)</a></li>
<li class="toctree-l2"><a class="reference internal" href="cmdline.html#get-experiment-accessions-for-a-project-srp-srx">Get experiment accessions for a project (SRP =&gt; SRX)</a></li>
<li class="toctree-l2"><a class="reference internal" href="cmdline.html#get-sample-accessions-for-a-project-srp-srs">Get sample accessions for a project (SRP =&gt; SRS)</a></li>
<li class="toctree-l2"><a class="reference internal" href="cmdline.html#get-run-accessions-for-experiments-srx-srr">Get run accessions for experiments (SRX =&gt; SRR)</a></li>
<li class="toctree-l2"><a class="reference internal" href="cmdline.html#get-experiment-accessions-for-runs-srr-srx">Get experiment accessions for runs (SRR =&gt; SRX)</a></li>
<li class="toctree-l2"><a class="reference internal" href="cmdline.html#get-geo-accessions-for-studies-srp-gse">Get GEO accessions for studies (SRP =&gt; GSE)</a></li>
<li class="toctree-l2"><a class="reference internal" href="cmdline.html#get-sra-accessions-for-geo-studies-gse-srp">Get SRA accessions for GEO studies (GSE =&gt; SRP)</a></li>
<li class="toctree-l2"><a class="reference internal" href="cmdline.html#searching-sra">Searching SRA</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="python-api-usage.html">Python API Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="python-api-usage.html#use-case-1-fetch-the-metadata-table-sra-runtable">Use Case 1: Fetch the metadata table (SRA-runtable)</a></li>
<li class="toctree-l2"><a class="reference internal" href="python-api-usage.html#use-case-2-downloading-an-entire-project-arranged-experiment-wise">Use Case 2: Downloading an entire project arranged experiment wise</a></li>
<li class="toctree-l2"><a class="reference internal" href="python-api-usage.html#use-case-3-downloading-a-subset-of-experiments">Use Case 3: Downloading a subset of experiments</a></li>
<li class="toctree-l2"><a class="reference internal" href="python-api-usage.html#use-case-4-getting-cell-type-treatment-information-from-sample-attributes">Use Case 4: Getting cell-type/treatment information from sample_attributes</a></li>
<li class="toctree-l2"><a class="reference internal" href="python-api-usage.html#use-case-5-searching-for-datasets">Use Case 5: Searching for datasets</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="usage_scenarios.html">Usage Scenarios</a><ul>
<li class="toctree-l2"><a class="reference internal" href="usage_scenarios.html#scenario-1">Scenario 1</a></li>
<li class="toctree-l2"><a class="reference internal" href="usage_scenarios.html#scenario-2">Scenario 2</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="commands.html">pysradb API Documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="commands/download.html">download</a></li>
<li class="toctree-l2"><a class="reference internal" href="commands/metadata.html">metadata</a></li>
<li class="toctree-l2"><a class="reference internal" href="commands/search.html">search</a></li>
<li class="toctree-l2"><a class="reference internal" href="commands/gse-to-gsm.html">gse-to-gsm</a></li>
<li class="toctree-l2"><a class="reference internal" href="commands/gse-to-srp.html">gse-to-srp</a></li>
<li class="toctree-l2"><a class="reference internal" href="commands/srp-to-gse.html">srp-to-gse</a></li>
<li class="toctree-l2"><a class="reference internal" href="commands/srp-to-srr.html">srp-to-srr</a></li>
<li class="toctree-l2"><a class="reference internal" href="commands/srp-to-srs.html">srp-to-srs</a></li>
<li class="toctree-l2"><a class="reference internal" href="commands/srp-to-srx.html">srp-to-srx</a></li>
<li class="toctree-l2"><a class="reference internal" href="commands/srr-to-srs.html">srr-to-srs</a></li>
<li class="toctree-l2"><a class="reference internal" href="commands/srr-to-srx.html">srr-to-srx</a></li>
<li class="toctree-l2"><a class="reference internal" href="commands/srs-to-srx.html">srs-to-srx</a></li>
<li class="toctree-l2"><a class="reference internal" href="commands/srx-to-srr.html">srx-to-srr</a></li>
<li class="toctree-l2"><a class="reference internal" href="commands/srx-to-srs.html">srx-to-srs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">pysradb modules</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pysradb.html">pysradb package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a><ul>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#types-of-contributions">Types of Contributions</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#get-started">Get Started!</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#pull-request-guidelines">Pull Request Guidelines</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#tips">Tips</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#deploying">Deploying</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a><ul>
<li class="toctree-l2"><a class="reference internal" href="authors.html#development-lead">Development Lead</a></li>
<li class="toctree-l2"><a class="reference internal" href="authors.html#contributors">Contributors</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a><ul>
<li class="toctree-l2"><a class="reference internal" href="history.html#id1">0.10.3 (03-26-2020)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id2">0.10.2 (02-05-2020)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id3">0.10.1 (02-04-2020)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id4">0.10.0 (01-31-2020)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id5">0.9.9 (01-15-2020)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id6">0.9.7 (01-20-2020)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id7">0.9.6 (07-20-2019)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id8">0.9.0 (02-27-2019)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id9">0.8.0 (02-26-2019)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id11">0.7.1 (02-18-2019)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id13">0.7.0 (02-08-2019)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id15">0.6.0 (12-25-2018)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id18">0.5.0 (12-24-2018)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id20">0.4.2 (12-16-2018)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id22">0.4.0 (12-12-2018)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id24">0.3.0 (12-05-2018)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id26">0.2.2 (12-03-2018)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id28">0.2.0 (12-03-2018)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id31">0.1.0 (12-01-2018)</a></li>
</ul>
</li>
</ul>
</div>
</div>
<hr class="docutils" />
<div class="section" id="publication">
<h2>Publication<a class="headerlink" href="#publication" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><p><a class="reference external" href="https://f1000research.com/articles/8-532/v1">pysradb: A Python package to query next-generation sequencing metadata and data from NCBI Sequence Read Archive</a></p>
<p>Presentation slides from BOSC (ISMB-ECCB) 2019: <a class="reference external" href="https://f1000research.com/slides/8-1183">https://f1000research.com/slides/8-1183</a></p>
</div></blockquote>
</div>
<hr class="docutils" />
<div class="section" id="citation">
<h2>Citation<a class="headerlink" href="#citation" title="Permalink to this headline">¶</a></h2>
<p>Choudhary, Saket. “pysradb: A Python Package to Query next-Generation Sequencing Metadata and Data from NCBI Sequence Read Archive.” F1000Research, vol. 8, F1000 (Faculty of 1000 Ltd), Apr. 2019, p. 532 (<a class="reference external" href="https://f1000research.com/articles/8-532/v1">https://f1000research.com/articles/8-532/v1</a>)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nd">@article</span><span class="p">{</span><span class="n">Choudhary2019</span><span class="p">,</span>
<span class="n">doi</span> <span class="o">=</span> <span class="p">{</span><span class="mf">10.12688</span><span class="o">/</span><span class="n">f1000research</span><span class="o">.</span><span class="mf">18676.1</span><span class="p">},</span>
<span class="n">url</span> <span class="o">=</span> <span class="p">{</span><span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">doi</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="mf">10.12688</span><span class="o">/</span><span class="n">f1000research</span><span class="o">.</span><span class="mf">18676.1</span><span class="p">},</span>
<span class="n">year</span> <span class="o">=</span> <span class="p">{</span><span class="mi">2019</span><span class="p">},</span>
<span class="n">month</span> <span class="o">=</span> <span class="n">apr</span><span class="p">,</span>
<span class="n">publisher</span> <span class="o">=</span> <span class="p">{</span><span class="n">F1000</span> <span class="p">(</span><span class="n">Faculty</span> <span class="n">of</span> <span class="mi">1000</span> <span class="n">Ltd</span><span class="p">)},</span>
<span class="n">volume</span> <span class="o">=</span> <span class="p">{</span><span class="mi">8</span><span class="p">},</span>
<span class="n">pages</span> <span class="o">=</span> <span class="p">{</span><span class="mi">532</span><span class="p">},</span>
<span class="n">author</span> <span class="o">=</span> <span class="p">{</span><span class="n">Saket</span> <span class="n">Choudhary</span><span class="p">},</span>
<span class="n">title</span> <span class="o">=</span> <span class="p">{</span><span class="n">pysradb</span><span class="p">:</span> <span class="n">A</span> <span class="p">{</span><span class="n">P</span><span class="p">}</span><span class="n">ython</span> <span class="n">package</span> <span class="n">to</span> <span class="n">query</span> <span class="nb">next</span><span class="o">-</span><span class="n">generation</span> <span class="n">sequencing</span> <span class="n">metadata</span> <span class="ow">and</span> <span class="n">data</span> <span class="kn">from</span> <span class="p">{</span><span class="n">NCBI</span><span class="p">}</span> <span class="p">{</span><span class="n">S</span><span class="p">}</span><span class="n">equence</span> <span class="p">{</span><span class="n">R</span><span class="p">}</span><span class="n">ead</span> <span class="p">{</span><span class="n">A</span><span class="p">}</span><span class="n">rchive</span><span class="p">},</span>
<span class="n">journal</span> <span class="o">=</span> <span class="p">{</span><span class="n">F1000Research</span><span class="p">}</span>
<span class="p">}</span>
</pre></div>
</div>
<p>Zenodo archive: <a class="reference external" href="https://zenodo.org/badge/latestdoi/159590788">https://zenodo.org/badge/latestdoi/159590788</a></p>
<p>Zenodo DOI: 10.5281/zenodo.2306881</p>
</div>
<hr class="docutils" />
<div class="section" id="questions">
<h2>Questions?<a class="headerlink" href="#questions" title="Permalink to this headline">¶</a></h2>
<p>Join our <a class="reference external" href="https://join.slack.com/t/pysradb/shared_invite/zt-f01jndpy-KflPu3Be5Aq3FzRh5wj1Ug">Slack Channel</a> or open an <a class="reference external" href="https://github.com/saketkc/pysradb/issues">issue</a>.</p>
<ul class="simple">
<li><p>Free software: BSD license</p></li>
<li><p>Documentation: <a class="reference external" href="https://saketkc.github.io/pysradb">https://saketkc.github.io/pysradb</a></p></li>
</ul>
</div>
<div class="section" id="indices-and-tables">
<h2>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="commands/search.html#search"><span class="std std-ref">search</span></a></p></li>
</ul>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="#">pysradb</a></h1>








<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="cmdline.html">CLI Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="python-api-usage.html">Python API Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage_scenarios.html">Usage Scenarios</a></li>
<li class="toctree-l1"><a class="reference internal" href="commands.html">pysradb API Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">pysradb modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
      <li>Next: <a href="quickstart.html" title="next chapter">Quickstart</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2018, Saket Choudhary.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>